### ISSUE TITLE
multimask_output with SAM-HQ

### ISSUE THREAD

**Issue by danigarciaoca** [SysCV/sam-hq#19]
Hi everyone!

I was comparing the implementation of original `mask_decoder.py` and improved `mask_decoder_hq.py` and I have found a difference when using `multimask_output` flag.

In the original implementation ([mask_decoder.py](https://github.com/facebookresearch/segment-anything/blob/main/segment_anything/modeling/mask_decoder.py#L102)), when using `multimask_output` the output was a mask with size 1, 3, 256, 256 and iou_pred with size 1, 3.
However, with the new implemententation ([mask_decoder_hq.py](https://github.com/SysCV/sam-hq/blob/main/segment_anything/modeling/mask_decoder_hq.py#L136)), when using `multimask_output` the output is a mask with size 1, 1, 256, 256 and iou_pred with size 1,1. Furthermore, this mask is the one associated with the maximum iou_pred (that is, the mask with the maximum "detection confidence").

In summary:
- mask_decoder.py and `multimask_output==True`
    - masks: Tensor (1, **3**, 256, 256)
    - iou_pred = Tensor (1, **3**)
- mask_decoder_hq.py and `multimask_output==True`
    - masks_sam: Tensor (1, **1**, 256, 256)
    - iou_pred = Tensor (1, **1**)

Is there any explanation for this implementation? When `multimask_output==True`, SAM used to return 3 masks but SAM-HQ now just returns 1.

In the new implementation, then masks_sam is added to masks_hq to obtain final masks (`masks = masks_sam + masks_hq`). Couldn't this be done over the 3 obtained masks instead of selecting just the one that maximizes iou_prob?

Thanks in advance and great work @lkeab!

**Comment by lkeab** [comment]
Hi danigarciaoca, thanks for watching our work. SAM-HQ is currently trained to optimize the best prediction result of SAM, which is the only mask outputted in single mask mode and the mask with max scoring in multi-mask mode. It could be done over the 3 obtained masks but the visual results may degrade.

**Comment by danigarciaoca** [comment]
Okey, it make sense. And is there any recommendation about just using multimask = False or True? Or the same criteria as with original SAM implementation (that is, multimask=True when ambigous input is provided and multimask=False otherwise)

**Comment by lkeab** [comment]
Yes. we suggest to follow SAM's implementation on using it.

**Comment by danigarciaoca** [comment]
Thanks! I'm closing this issue :smile:
